Implement conflict detection in strategic merge patch

jackgr opened the pull request:
Per #13576.

I undertook this work, since #13576 was fresh in my mind, thinking that it would be needed for apply --force, per #16569. However, looking at the discussion there, I now see that the expected behavior of the --force flag is to delete and recreate.

Bottom line, it's nice to have an implementation of #13576, but it doesn't get us any closer to #16569.

cc @bgrant0607, @ghodss, @janetkuo, @jlowdermilk

Commit:
Implement conflict detection in strategic merge patch.

bgrant0607 commented:
Thanks, @jackgr. Could you please describe what this does -- the user experience? 

We should update the docs, also, though I believe your previous doc PR hasn't merged yet.

jackgr commented:
@bgrant0607 I've updated the comment on CreateThreeWayMergePatch to be more explicit. See [line 1129](https://github.com/jackgr/kubernetes/blob/force_apply/pkg/util/strategicpatch/patch.go#L1129).

Agreed that the documentation for `kubectl apply` needs to be updated, but that change can be made in the next PR, when `kubectl apply` is modified to use this functionality.

Speaking of which, how should it be modified? 

#16569 suggests that `--force` should delete and recreate the object. However, `kubectl replace` does exactly that.

We could let the caller use `kubectl replace`, instead, or we could add that functionality to `kubectl apply` with a `--replace` flag. Either way, we could then use `kubectl apply --force` to call this method with the force parameter set to `true`.

jackgr commented:
Note: test coverage is 80.3% according to go test.

bgrant0607 commented:
To clarify: If a conflict is detected, this will make kubectl apply return an error rather than silently allow the conflict to persist, unless --force is specified, in which case the conflict will be clobbered by the patch, assuming that the field is updatable, correct?

Does apply aggregate errors, or does it stop on the first error?

Re. apply --force vs. replace --force: apply is "make it so". apply --force is "make it so, dammit!" I don't want apply users to have to use other forms of updates in different scenarios. apply should be smart enough to figure it out. For instance, it shouldn't delete and re-create unless that's necessary. Also, I could also imagine apply doing an export (#13038), merge/patch on client side, and then re-create, thus preserving more of the current operational state than replace.

jackgr commented:
@bgrant0607 Yes, that's correct. It stops on the first error and returns an error that contains the conflicting configurations. Currently, it does not identify the specific conflict that caused the error, but leaves it to the caller to visually compare the two configurations.

jackgr commented:
@bgrant0607 To clarify re: apply --force vs. replace --force, I was actually proposing apply --force vs. apply --replace, with the following interpretations:

* apply (make it so)
* apply --force (make it so, dammit)
* apply --replace (make it so, even if you have to delete and recreate the object)

The purpose of this PR is to allow the construction of an apply --force option that overwrites conflicting changes. I suppose we could try that first, and then delete and recreate the object if it fails, but that seems a bit draconian, thus the suggestion that we provide an apply --replace option, which deletes and recreates the object if and only if the overwrite fails.

bgrant0607 commented:
I'd like --force to have a consistent meaning.

How about --ignore-conflicts for (2) and --force for (3), which would also imply (2).

jackgr commented:
Great suggestion. However, `--ignore-conflicts` is a bit ambiguous. Does it mean that the conflicting values are not applied, or that they overwrite the current values. How about `--overwrite` to be more explicit?

deads2k commented on pkg/util/strategicpatch/patch.go:
```diff
@@ -963,13 +972,170 @@ func HasConflicts(left, right interface{}) (bool, error) {
 	}
 }
 
+// MergingMapsHaveConflicts returns true if the left and right JSON interface
```
This comment looks almost identical to https://github.com/kubernetes/kubernetes/blob/master/pkg/util/strategicpatch/patch.go#L930.  What's the difference?

jackgr replied:
This method supports merging lists (i.e., strategic merge path semantics), while the other one doesn't (i.e., json merge patch semantics). Will update to make the distinction clear.

deads2k replied:
> This method supports merging lists (i.e., strategic merge path semantics), while the other one doesn't (i.e., json merge patch semantics). Will update to make the distinction clear.

If this is a superset, should you simply replace the existing one?

jackgr replied:
The implementation started out that way, but it quickly became unmanageable. Also, the existing method is being used in its current form, so going that route meant create a stub that offered the old API over the new one, passing in a nil structure for type information, and then plumbing that through the implementation. In the end, it was simpler and cleaner to leave the existing method unchanged, and to write the new one straightforwardly using strategic merge patch semantics.

deads2k commented on pkg/util/strategicpatch/patch.go:
```diff
@@ -963,13 +972,170 @@ func HasConflicts(left, right interface{}) (bool, error) {
+			leftMarker, okLeft := leftType[directiveMarker]
+			rightMarker, okRight := rightType[directiveMarker]
+			if okLeft || okRight {
+				if okLeft != okRight || leftMarker != rightMarker {
```
if `okLeft != okRight`, then at least one is true, so this `if` doesn't need to be nested inside another `if`.  Can you promote it out so that the blocks are shallow and short?

jackgr replied:
Yes.

deads2k commented on pkg/util/strategicpatch/patch.go:
```diff
@@ -963,13 +972,170 @@ func HasConflicts(left, right interface{}) (bool, error) {
+		case map[string]interface{}:
+			leftMarker, okLeft := leftType[directiveMarker]
+			rightMarker, okRight := rightType[directiveMarker]
+			if okLeft || okRight {
```
If I'm reading this right, it goes like this:

 1. if leftMarker != rightMarker, return conflict
 1. if okLeft != okRight, return no conflict
 1. otherwise, check the maps.

jackgr replied:
I think of it like this:

1. if one or the other has a directive marker, then we need to consider that before looking at the individual keys, since a directive operates on the whole map.
2. if one has a directive marker and the other doesn't, then we have a conflict, since one is deleting or replacing the whole map, and the other is doing things to individual keys.
3. if they both have markers, but they are not the same directive, then we have a conflict because they're doing different things to the map.
4. if they both have the same directive, then there's no conflict here, and we're done looking at this map.
5. otherwise, check the individual keys.

This seems obtuse enough to warrant a comment.

deads2k commented on pkg/util/strategicpatch/patch.go:
```diff
@@ -66,8 +69,8 @@ type errConflict struct {
-func newErrConflict(patch, current []byte) errConflict {
-	s := fmt.Sprintf("patch:\n%s\nconflicts with current:\n%s\n", patch, current)
+func newErrConflict(patch, current string) errConflict {
+	s := fmt.Sprintf("patch:\n%s\nconflicts with automated changes:\n%s\n", patch, current)
```
Why is it valid to assume that this was changed by an automated process?  It could have been any client that didn't know to update the magic annotation, right?

deads2k replied:
> Why is it valid to assume that this was changed by an automated process? It could have been any client that didn't know to update the magic annotation, right?

In fact, even kube clients of skewed versions will probably get this wrong because of version and codec changes, right?

jackgr replied:
Re: the assumption that it was an automated process, we require that clients know about and use the annotation, automated processes do not. No magic there. It's part of the interface, and documented as such.

Re: version skew, good question. Wrote up #16543 and then did some investigation. See the results there. Bottom line, we need to a version check that returns an error if the modified version is older than the original version.

deads2k commented on pkg/util/strategicpatch/patch.go:
```diff
@@ -963,13 +972,170 @@ func HasConflicts(left, right interface{}) (bool, error) {
+// than from original to current. In other words, a conflict occurs if modified changes any key
+// in a way that is different from how it is changed in current (e.g., deleting it, changing its
+// value).
 func CreateThreeWayMergePatch(original, modified, current []byte, dataStruct interface{}, force bool, fns ...PreconditionFunc) ([]byte, error) {
```
Looking at how this is called, it looks like the `original` is retrieved and then isn't necessarily getting converted to the correct version for the comparison to `modified` and `current`.  Am I just missing it?

jackgr replied:
Well, that would be a comment for the call site, not for this method, since this method has no knowledge of who's calling it, or how they obtained the byte slices.

deads2k commented:
I'd like to understand the fundamentals of how the data is being stored and transformed to properly span versions of differently skewed clients that may even have different versions of the same scheme before I try to review something building on top of it.

There's a significant amount of code to make this work, so its entirely possible that I'm simply overlooking it.

@liggitt you looked into this recently.  If you have pointers to where this is handled it would help.

liggitt commented:
The only bits I looked at were the implementations of `StrategicMergePatch`, etc, in pkg/util/strategicpatch, but those are operating on raw JSON at that point, so I'm not sure how all the callers of those compute their JSON byte arrays

liggitt commented:
What I looked at is fixed in https://github.com/kubernetes/kubernetes/pull/16964

deads2k commented:
> I'd like to understand the fundamentals of how the data is being stored and transformed to properly span versions of differently skewed clients that may even have different versions of the same scheme before I try to review something building on top of it.

Looks like there's already an issue describing this here: https://github.com/kubernetes/kubernetes/issues/16543

bgrant0607 commented:
I'm fine with --overwrite, @jackgr

bgrant0607 commented:
Is this stuck?

jackgr commented:
This PR is ready to go, and has been for some time. It implements the --overwrite functionality at the level of strategic merge patch. In other words, it allows the caller of StrategicMergePatch to overwrite conflicts by passing `true` for the `force` argument, which might arguably have been better named `overwrite`.

To finish out kubectl apply --overwrite, which uses it to support overwrite, and kubectl apply --force, which deletes and then recreates the object if and only if overwrite fails, I need to implement #17238. That work depends on this PR.

deads2k commented:
> Is this stuck?

I was hoping for a cleaner resolution in #16543  than

> That said, if we can't find a clean and simple way to use their historical context when the version changes, then maybe it is the best experience we can offer. At this point, I'm tempted to do just that, since this feature is dragging out a long time.

but if that's as good as it gets (feature is kind of borked across API versions and is heavily client-logic dependent), then I'll go back through this.

liggitt commented:
Did this PR actually get updated to indicate conflicts if apiVersion or kind doesn't match? Didn't see it in a quick scan through

deads2k commented on pkg/util/strategicpatch/patch.go:
```diff
@@ -963,13 +973,185 @@ func HasConflicts(left, right interface{}) (bool, error) {
+
+				// if they both have the same directive,
+				// then there's no conflict here, and we're done looking at this map.
+				return false, nil
```
Not recalling the exact format of a strategic merge patch, if I have a directive, can't I still have elements in that map that need to be compared?

deads2k commented:
> Did this PR actually get updated to indicate conflicts if apiVersion or kind doesn't match? Didn't see it in a quick scan through

If those change, the update will eventually fail.

deads2k commented:
The implementation looks alright, just one more question about the conflict detection.  I'm still concerned that the overall approach seems fragile in the general case.

jackgr commented:
It's not the job of this code to check the API version as noted above.
That's the job of kubectl apply. This code knows nothing about Kubernetes
API, codecs, etc.

When this PR is merged, I will implement #17238 (
https://github.com/kubernetes/kubernetes/issues/17238) to add the
--overwrite and --force flags. At that time, I will add the check for
version skew between the annotation and the input, and discard the
annotation of they differ.

Sorry, I thought this had been clearly stated already.

On Friday, November 20, 2015, David Eads <notifications@github.com> wrote:

> The implementation looks alright, just one more question about the
> conflict detection. I'm still concerned that the overall approach seems
> fragile in the general case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16980#issuecomment-158409503>
> .
>


-- 
Jack Greenfield | Google

jackgr commented:
It's not the job of this code to check the API version as noted above. That's the job of kubectl apply. This code knows nothing about Kubernetes API, codecs, etc.

When this PR is merged, I will implement #17238 to add the --overwrite and --force flags. At that time, I will add the check for version skew between the annotation and the input, and discard the annotation of they differ.

Sorry, I thought this had been clearly stated already.

jackgr commented:
Rebased and renamed the `force` argument to `overwrite` to align terminology with the command line flags. @bgrant0607 this PR is ready to merge, and has been ready for some time. As noted above, it's not the job of StrategicMergePatch or CreateStrategicMergePatch to detect version skew, since they don't know anything about Kubernetes API, codecs, etc. It therefore shouldn't be waiting on concerns about version skew. Those concerns should be addressed by the kubectl apply command, which is the subject of #16569 and #17238.

deads2k commented:
> @bgrant0607 this PR is ready to merge

I'm still wondering about https://github.com/kubernetes/kubernetes/pull/16980#discussion_r45469894.  The current code would look and say, "oh, you're both doing replaces, no conflict" instead of checking to make sure the replaces don't conflict, right?

jackgr commented:
The current code doesn't understand versions. It detects conflicts
purely on the basis of different changes to the same properties. It's up to
the code that calls it, in kubectl apply, to deal with version skew, which
it will handle by discarding the annotation when the version changes.

On Monday, November 23, 2015, David Eads <notifications@github.com> wrote:

> @bgrant0607 <https://github.com/bgrant0607> this PR is ready to merge
>
> I'm still wondering about #16980 (comment)
> <https://github.com/kubernetes/kubernetes/pull/16980#discussion_r45469894>.
> The current code would look and say, "oh, you're both doing replaces, no
> conflict" instead of checking to make sure the replaces don't conflict,
> right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16980#issuecomment-158924859>
> .
>


-- 
Jack Greenfield | Google

liggitt commented:
I think https://github.com/kubernetes/kubernetes/pull/16980#discussion-diff-45469894 was about generic merges on maps, not related to versions

deads2k commented:
> I'm still wondering about #16980 (comment). The current code would look and say, "oh, you're both doing replaces, no conflict" instead of checking to make sure the replaces don't conflict, right?

bump.  I'm still unclear on why its doing that.  It seems to conflict with the comment too.

jackgr commented:
Perhaps the best way to resolve these questions quickly is with a video
conference. Please send me an email so we can make arrangements offline
from this issue. Same user name at Google.

On Monday, November 30, 2015, David Eads <notifications@github.com> wrote:

> I'm still wondering about #16980
> <https://github.com/kubernetes/kubernetes/pull/16980> (comment). The
> current code would look and say, "oh, you're both doing replaces, no
> conflict" instead of checking to make sure the replaces don't conflict,
> right?
>
> bump. I'm still unclear on why its doing that. It seems to conflict with
> the comment too.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16980#issuecomment-160647601>
> .
>


-- 
Jack Greenfield | Google

deads2k commented:
> Perhaps the best way to resolve these questions quickly is with a video
> conference. Please send me an email so we can make arrangements offline
> from this issue. Same user name at Google.

Sent an email this morning, didn't see anything back.

The only thing holding this up is that the code here: https://github.com/kubernetes/kubernetes/pull/16980/files#diff-3b6b2da9a290774f14d23ece307b65baR1019 doesn't match the comment here: https://github.com/kubernetes/kubernetes/pull/16980/files#diff-3b6b2da9a290774f14d23ece307b65baR1001 (see existing comments in this thread).  It looks like the code as written would look and say, "oh, you're both doing replaces, no conflict" instead of checking to make sure the replaces don't conflict.

Concerns about the stability and predictability of `apply` can be discussed in the issues we already have open.

jackgr commented:
I see. Thanks for clarifying. Yes, you're quite right. It should go on to check the individual keys.

jackgr commented:
Missing commit. Fixed now.

deads2k commented:
lgtm.

deads2k commented:
@jackgr looks like this hasn't merged because of red tests.

jackgr commented:
Thanks for the heads up. Will take a look.

gmarek commented:
@k8s-bot test this

jackgr commented:
@deads2k Silly mistake. Forgot to change the flag value passed to CreateStrategicMergePatch in pkg/kubectl/cmd/apply.go from false to true. Now that we detect conflict, we need to force overwrite to get the same behavior as before. When #17238 is implemented, it will set the flag value according to the command line flags.

deads2k commented:
@k8s-bot test this please

@jackgr if this one doesn't pass, ptal.

jackgr commented:
@deads2k AFAICT all of the issues are related to the test infrastructure. I don't see any indication of a test failure in strategic merge patch. Am I missing someting?

deads2k commented:
> @deads2k AFAICT all of the issues are related to the test infrastructure. I don't see any indication of a test failure in strategic merge patch. Am I missing someting?

There have been a couple different ones.  Just figured I'd let you know that its still sitting here unhappily.

jackgr commented:
Yes. Thanks. Sigh.

jlowdermilk commented:
@k8s-bot e2e-test this

Submit queue is backed up due to broken infra. I'll manually merge this once it's green.

jackgr commented:
Thanks, @jlowdermilk.

wojtek-t commented:
@k8s-bot test this please

wojtek-t commented:
@k8s-bot test this please

wojtek-t commented:
@k8s-bot test this please

wojtek-t commented:
@k8s-bot unit test this please

spxtr commented:
@k8s-bot unit test this please

wojtek-t commented:
@k8s-bot unit test this please

wojtek-t commented:
@k8s-bot e2e test this please

wojtek-t commented:
@k8s-bot unit test this please

wojtek-t commented:
unit test failure is: #18929

@k8s-bot unit test this please
